### ISSUE TITLE
LightOccluder2D polygons cause editor slowdowns

### ISSUE THREAD

**Issue by HaSa1002** [godotengine/godot#52703]
### Godot version

3.3.3.stable

### System information

Windows 10, GLES3, Ryzen 7 3700X, 32GB RAM

### Issue description

LightOccluder2D polygons slow the editor down even with small polygons. (And my computer is not a potato)

https://user-images.githubusercontent.com/14185889/133438422-9841b2b7-f798-4b0b-87f7-8f97b6a3f2c5.mp4

### Steps to reproduce

1. Open `gear2.tscn`
2. Select `LightOccluder2D` (<- slowdown starts here)
3. Pan the canvas

### Minimal reproduction project

[Slow LightOccluder.zip](https://github.com/godotengine/godot/files/7170268/Slow.LightOccluder.zip)

**Comment by HaSa1002** [comment]
Yes, I can hook VS into Godot once I am done with my current task

**Comment by lawnjelly** [comment]
Could well be similar to #38543, possibly issuing a large number of drawcalls to draw the preview.

**Comment by akien-mga** [comment]
Could you check if #54377 fixes it? I tested locally and couldn't reproduce the slowdown before this PR either, so I can't confirm.

**Comment by HaSa1002** [comment]
Tested on `v3.5.beta.custom_build.4cae48d81` and issue is still noticable. Though not so extreme (just a feeling. May be incorrect)

**Comment by HaSa1002** [comment]
I can share the diagsession with someone if there is interest. I just don't want to post it here. I don't see anything suspicious. (OK my GPU get's one fire...)

**Comment by lawnjelly** [comment]
It's okay shouldn't need the project. I tried the min reproduction project again, and compiled it so the occluder debug draws even outside the editor, so I could diagnose the frame with the batching.

```

canvas_begin FRAME 554
items
	joined_item 1 refs, 
			batch R 0-1 [0 - 152] {255 255 255 255 }
	joined_item 1 refs, 
			batch P 0-138 {255 255 255 255 } MULTI
canvas_end

```
This is the output, so the debug drawing is the second, as a polygon all in one drawcall.
Looking at the calling code, this may be the same as before the PR above, so the drawcalls is not a problem, it's drawing everything in just 2 drawcall.
```
			PoolVector<Vector2> poly = occluder_polygon->get_polygon();

				if (poly.size()) {
					if (occluder_polygon->is_closed()) {
						Vector<Color> color;
						color.push_back(Color(0, 0, 0, 0.6));
						draw_polygon(Variant(poly), color);
					} else {
						int ps = poly.size();
						PoolVector<Vector2>::Read r = poly.read();
						for (int i = 0; i < ps - 1; i++) {
							draw_line(r[i], r[i + 1], Color(0, 0, 0, 0.6), 3);
						}
					}
				}
```
So not immediately sure why it might be slow. :thinking:

**Comment by HaSa1002** [comment]
Could that be a driver issue? I mean the 3060 TI should be able to draw this poly. On the other hand the core usage of main thread goes up to 100%

**Comment by lawnjelly** [comment]
Possibly, it's strange behaviour, if it's with the MRP. If you are able to use a CPU profiler, and you find a load of time is spent in the driver, then that sees like a culprit.

Some things you could try if that was the case:
* Turn off batching in the editor (`rendering/batching/options/use_batching_in_editor`) and restart. The legacy may use a different draw method (although I'm not sure it may be the same for polys, but worth a try)
* In `rendering/2d/openg` there are 4 settings for the dynamic buffer usage. You could try tweaking them for your system, you may have some joy.

**Comment by HaSa1002** [comment]
> load of time is spent in the driver, then that sees like a culprit.

Idk tbf. I could try to update my driver.

> Turn off batching in the editor (rendering/batching/options/use_batching_in_editor) and restart. The legacy may use a different draw method (although I'm not sure it may be the same for polys, but worth a try)

Did that. No change.

> In rendering/2d/openg there are 4 settings for the dynamic buffer usage. You could try tweaking them for your system, you may have some joy.

will try

**Comment by Penjami** [comment]
@HaSa1002 Did you end up finding what was the cause for the problem? I'm having similar performance issues with the LightOccluder2D. Hard to draw a polygon, when it gets super laggy after drawing about 15 vertices.

**Comment by Calinou** [comment]
This comment still applies, by the way:

> Can you try [running a C++ profiler](https://docs.godotengine.org/en/latest/development/cpp/using_cpp_profilers.html) to see where the slowdown comes from?

(Please post results of profiling here by taking screenshots.)

**Comment by Penjami** [comment]
Here is a couple of screenshots from running the profile for about a minute while modifying a LightOccluder2D polygon: 

With godot.windows.tools.64 filter:
![image](https://user-images.githubusercontent.com/25846175/212471427-c7f165b8-102e-43c5-a50b-524f86ec935c.png)

And without:
![image](https://user-images.githubusercontent.com/25846175/212471438-50625a08-e2d3-41fe-8d98-c4a8df203e3b.png)

**Comment by lawnjelly** [comment]
That does look like it's spending a lot of time in the driver (DrvValidateVersion seems prominent, and some sleeps and waits, which might be indicative of stalls).

**Comment by deyarmol** [comment]
I'm interested in claiming this as a first contribution (for a school assignment). Is that acceptable?

**Comment by lawnjelly** [comment]
There's no need to claim.
Whether this is suitable for a first issue I'm not absolutely sure (if I remember it correctly), but feel free to have an investigate.

**Comment by deyarmol** [comment]
I was discussing this with someone else and we found this issue that may be similar: https://github.com/opentk/opentk/issues/474. I'm trying to track down where the relevant OpenGL operations would be done in the code so I can check that directly (looking at logs hasn't been very informative so far) but I'm having trouble figuring out exactly where those are...

**Comment by deyarmol** [comment]
Yeah, unfortunately I think this is actually out of my depth and I should probably look for something else to work on. I'm sorry for the inconvenience!

**Comment by yuantianle** [comment]
To be honest, I did not see lagging on my end. Not sure if it is the hardware issue. I am using:
Windows 11
13th Gen Intel Core i9-13980
NVIDIA GeForce RTX 4070 laptop,
64GB RAM
